File: Source/Blazorise/Services/DocumentObserver.cs
#region Using directives
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Blazorise.Modules;
using Microsoft.JSInterop;
#endregion

namespace Blazorise;

/// <summary>
/// Default implementation of <see cref="IDocumentObserver"/>.
/// </summary>
public class DocumentObserver : IDocumentObserver
{
    #region Members

    private readonly IJSDocumentObserverModule jsDocumentObserverModule;

    private readonly Dictionary<string, DocumentObserverSubscription> subscriptions = new();

    private DotNetObjectReference<DocumentObserverAdapter> dotNetObjectRef;

    private Task initializationTask;

    private int nextSubscriptionId;

    #endregion

    #region Constructors

    /// <summary>
    /// Default constructor for <see cref="DocumentObserver"/>.
    /// </summary>
    /// <param name="jsDocumentObserverModule">The document observer JavaScript module.</param>
    public DocumentObserver( IJSDocumentObserverModule jsDocumentObserverModule )
    {
        this.jsDocumentObserverModule = jsDocumentObserverModule;
    }

    #endregion

    #region Methods

    /// <inheritdoc/>
    public Task EnsureInitializedAsync()
    {
        return initializationTask ??= InitializeAsync();
    }

    /// <inheritdoc/>
    public async ValueTask<IAsyncDisposable> Subscribe( DocumentObserverSubscription subscription )
    {
        if ( subscription is null )
            throw new ArgumentNullException( nameof( subscription ) );

        if ( subscription.EventTypes == DocumentEventTypes.None )
            throw new ArgumentException( "At least one document event type must be specified.", nameof( subscription ) );

        string subscriptionId = CreateSubscriptionId();

        try
        {
            subscriptions[subscriptionId] = subscription;

            await EnsureInitializedAsync();
            await jsDocumentObserverModule.AddSubscription( CreateJsSubscription( subscriptionId, subscription ) );
        }
        catch
        {
            subscriptions.Remove( subscriptionId );
            throw;
        }

        return new DocumentObserverSubscriptionHandle( this, subscriptionId );
    }

    /// <inheritdoc/>
    public async ValueTask CapturePointer( string ownerId, long pointerId )
    {
        await EnsureInitializedAsync();
        await jsDocumentObserverModule.CapturePointer( ownerId, pointerId );
    }

    /// <inheritdoc/>
    public async ValueTask ReleasePointer( string ownerId, long pointerId )
    {
        await EnsureInitializedAsync();
        await jsDocumentObserverModule.ReleasePointer( ownerId, pointerId );
    }

    /// <inheritdoc/>
    public async ValueTask DisposeAsync()
    {
        try
        {
            foreach ( string subscriptionId in subscriptions.Keys.ToArray() )
                await jsDocumentObserverModule.RemoveSubscription( subscriptionId );

            if ( jsDocumentObserverModule is IAsyncDisposable asyncDisposable )
                await asyncDisposable.DisposeAsync();
        }
        finally
        {
            subscriptions.Clear();
            dotNetObjectRef?.Dispose();
            dotNetObjectRef = null;
        }
    }

    internal async Task NotifyDocumentEvent( string subscriptionId, DocumentEventArgs eventArgs )
    {
        if ( !subscriptions.TryGetValue( subscriptionId, out DocumentObserverSubscription subscription )
             || subscription.Handler is null )
            return;

        await subscription.Handler( eventArgs );
    }

    private async Task InitializeAsync()
    {
        dotNetObjectRef ??= DotNetObjectReference.Create( new DocumentObserverAdapter( this ) );

        await jsDocumentObserverModule.Initialize( dotNetObjectRef );
    }

    private async ValueTask Unsubscribe( string subscriptionId )
    {
        if ( string.IsNullOrWhiteSpace( subscriptionId ) )
            return;

        if ( subscriptions.Remove( subscriptionId ) )
            await jsDocumentObserverModule.RemoveSubscription( subscriptionId );
    }

    private string CreateSubscriptionId()
    {
        nextSubscriptionId++;

        return $"b-document-observer-{nextSubscriptionId}";
    }

    private static DocumentObserverJsSubscription CreateJsSubscription( string subscriptionId, DocumentObserverSubscription subscription )
    {
        return new()
        {
            Id = subscriptionId,
            OwnerId = subscription.OwnerId,
            EventNames = GetEventNames( subscription.EventTypes ).ToArray(),
            Selector = subscription.Selector,
            ExcludeSelector = subscription.ExcludeSelector,
            Priority = subscription.Priority,
            Capture = subscription.Capture,
            PreventDefault = subscription.PreventDefault,
            StopPropagation = subscription.StopPropagation,
            Throttle = subscription.Throttle,
            IgnorePointerCapture = subscription.IgnorePointerCapture,
            DotNet = subscription.Handler is not null,
        };
    }

    private static IEnumerable<string> GetEventNames( DocumentEventTypes eventTypes )
    {
        if ( eventTypes.HasFlag( DocumentEventTypes.PointerDown ) )
            yield return "pointerdown";

        if ( eventTypes.HasFlag( DocumentEventTypes.PointerMove ) )
            yield return "pointermove";

        if ( eventTypes.HasFlag( DocumentEventTypes.PointerUp ) )
            yield return "pointerup";

        if ( eventTypes.HasFlag( DocumentEventTypes.PointerCancel ) )
            yield return "pointercancel";

        if ( eventTypes.HasFlag( DocumentEventTypes.MouseDown ) )
            yield return "mousedown";

        if ( eventTypes.HasFlag( DocumentEventTypes.MouseMove ) )
            yield return "mousemove";

        if ( eventTypes.HasFlag( DocumentEventTypes.MouseUp ) )
            yield return "mouseup";

        if ( eventTypes.HasFlag( DocumentEventTypes.TouchMove ) )
            yield return "touchmove";

        if ( eventTypes.HasFlag( DocumentEventTypes.TouchEnd ) )
            yield return "touchend";

        if ( eventTypes.HasFlag( DocumentEventTypes.TouchCancel ) )
            yield return "touchcancel";

        if ( eventTypes.HasFlag( DocumentEventTypes.Click ) )
            yield return "click";

        if ( eventTypes.HasFlag( DocumentEventTypes.DoubleClick ) )
            yield return "dblclick";

        if ( eventTypes.HasFlag( DocumentEventTypes.KeyDown ) )
            yield return "keydown";

        if ( eventTypes.HasFlag( DocumentEventTypes.KeyUp ) )
            yield return "keyup";

        if ( eventTypes.HasFlag( DocumentEventTypes.FocusIn ) )
            yield return "focusin";

        if ( eventTypes.HasFlag( DocumentEventTypes.FocusOut ) )
            yield return "focusout";

        if ( eventTypes.HasFlag( DocumentEventTypes.DragStart ) )
            yield return "dragstart";

        if ( eventTypes.HasFlag( DocumentEventTypes.DragOver ) )
            yield return "dragover";

        if ( eventTypes.HasFlag( DocumentEventTypes.DragEnd ) )
            yield return "dragend";

        if ( eventTypes.HasFlag( DocumentEventTypes.Drop ) )
            yield return "drop";

        if ( eventTypes.HasFlag( DocumentEventTypes.ContextMenu ) )
            yield return "contextmenu";

        if ( eventTypes.HasFlag( DocumentEventTypes.Blur ) )
            yield return "blur";
    }

    #endregion

    #region Classes

    private sealed class DocumentObserverSubscriptionHandle : IAsyncDisposable
    {
        private readonly DocumentObserver documentObserver;

        private readonly string subscriptionId;

        private bool disposed;

        public DocumentObserverSubscriptionHandle( DocumentObserver documentObserver, string subscriptionId )
        {
            this.documentObserver = documentObserver;
            this.subscriptionId = subscriptionId;
        }

        public async ValueTask DisposeAsync()
        {
            if ( disposed )
                return;

            disposed = true;

            await documentObserver.Unsubscribe( subscriptionId );
        }
    }

    #endregion
}

File: Source/Blazorise/wwwroot/documentObserver.js
const documentObserver = createDocumentObserver();

globalThis.Blazorise ??= {};
globalThis.Blazorise.documentObserver = documentObserver;

export function initialize(dotNetReference) {
    documentObserver.initialize(dotNetReference);
}

export function addSubscription(subscription) {
    documentObserver.subscribe(subscription);
}

export function removeSubscription(subscriptionId) {
    documentObserver.unsubscribe(subscriptionId);
}

export function capturePointer(ownerId, pointerId) {
    documentObserver.capturePointer(ownerId, pointerId);
}

export function releasePointer(ownerId, pointerId) {
    documentObserver.releasePointer(ownerId, pointerId);
}

function createDocumentObserver() {
    const subscriptions = new Map();
    const listeners = new Map();
    const pointerCaptures = new Map();
    const pendingAnimationFrameEvents = new Map();
    let dotNetReference = null;
    let nextContextElementId = 0;

    return {
        initialize,
        subscribe,
        unsubscribe,
        capturePointer,
        releasePointer,
    };

    function initialize(reference) {
        dotNetReference = reference;
    }

    function subscribe(subscription) {
        if (!subscription?.id || !Array.isArray(subscription.eventNames)) {
            return null;
        }

        unsubscribe(subscription.id);
        subscriptions.set(subscription.id, normalizeSubscription(subscription));
        updateListeners();

        return subscription.id;
    }

    function unsubscribe(subscriptionId) {
        if (!subscriptionId || !subscriptions.delete(subscriptionId)) {
            return;
        }

        updateListeners();
    }

    function capturePointer(ownerId, pointerId) {
        if (!ownerId || pointerId == null) {
            return;
        }

        pointerCaptures.set(String(pointerId), ownerId);
    }

    function releasePointer(ownerId, pointerId) {
        if (pointerId == null) {
            return;
        }

        const pointerKey = String(pointerId);

        if (!ownerId || pointerCaptures.get(pointerKey) === ownerId) {
            pointerCaptures.delete(pointerKey);
        }
    }

    function normalizeSubscription(subscription) {
        return {
            ...subscription,
            eventNames: subscription.eventNames.filter(Boolean),
            priority: Number.isFinite(subscription.priority) ? subscription.priority : 0,
            capture: subscription.capture !== false,
            preventDefault: subscription.preventDefault === true,
            stopPropagation: subscription.stopPropagation === true,
            throttle: subscription.throttle === true,
            ignorePointerCapture: subscription.ignorePointerCapture === true,
            dotNet: subscription.dotNet === true,
        };
    }

    function updateListeners() {
        const requiredKeys = new Set();

        for (const subscription of subscriptions.values()) {
            for (const eventName of subscription.eventNames) {
                for (const targetName of getListenerTargetNames(eventName)) {
                    requiredKeys.add(createListenerKey(eventName, subscription.capture, targetName));
                }
            }
        }

        for (const [key, listener] of listeners) {
            if (!requiredKeys.has(key)) {
                listener.target.removeEventListener(listener.eventName, listener.handler, listener.capture);
                listeners.delete(key);
            }
        }

        for (const key of requiredKeys) {
            if (listeners.has(key)) {
                continue;
            }

            const [eventName, captureValue, targetName] = key.split(":");
            const capture = captureValue === "1";
            const target = getListenerTarget(targetName);
            const handler = event => dispatch(event, capture);

            target.addEventListener(eventName, handler, { capture, passive: false });
            listeners.set(key, { eventName, capture, target, handler });
        }
    }

    function dispatch(event, capture) {
        const candidates = getCandidates(event, capture);

        for (const subscription of candidates) {
            if (!matchesPointerCapture(subscription, event) || !matchesTarget(subscription, event)) {
                continue;
            }

            if (subscription.preventDefault && typeof event.preventDefault === "function") {
                event.preventDefault();
            }

            if (subscription.stopPropagation && typeof event.stopPropagation === "function") {
                event.stopPropagation();
            }

            if (subscription.throttle) {
                dispatchThrottled(subscription, event);
            }
            else {
                invokeSubscription(subscription, event);
            }
        }
    }

    function getCandidates(event, capture) {
        return Array.from(subscriptions.values())
            .filter(subscription => subscription.capture === capture && subscription.eventNames.includes(event.type))
            .sort((a, b) => b.priority - a.priority);
    }

    function matchesPointerCapture(subscription, event) {
        if (subscription.ignorePointerCapture || event.pointerId == null || event.type === "pointerdown") {
            return true;
        }

        const ownerId = pointerCaptures.get(String(event.pointerId));

        return !ownerId || ownerId === subscription.ownerId;
    }

    function matchesTarget(subscription, event) {
        const target = event.target;

        if (!(target instanceof Element)) {
            return !subscription.selector && !subscription.excludeSelector;
        }

        if (subscription.excludeSelector && target.closest(subscription.excludeSelector)) {
            return false;
        }

        return !subscription.selector || !!target.closest(subscription.selector);
    }

    function dispatchThrottled(subscription, event) {
        const key = subscription.id;
        const current = pendingAnimationFrameEvents.get(key);

        if (current) {
            current.event = event;
            return;
        }

        const pending = { event };
        pendingAnimationFrameEvents.set(key, pending);

        requestAnimationFrame(() => {
            pendingAnimationFrameEvents.delete(key);
            invokeSubscription(subscription, pending.event);
        });
    }

    function invokeSubscription(subscription, event) {
        const args = createDocumentEventArgs(subscription, event);

        if (typeof subscription.handler === "function") {
            subscription.handler(event, args);
        }

        if (subscription.dotNet && dotNetReference) {
            dotNetReference.invokeMethodAsync("NotifyDocumentEvent", subscription.id, args);
        }
    }

    function createDocumentEventArgs(subscription, event) {
        const target = event.target instanceof Element ? event.target : null;
        const matched = subscription.selector && target ? target.closest(subscription.selector) : null;
        const contextElement = matched ?? target;
        const point = getEventPoint(event);

        return {
            type: getDocumentEventType(event.type),
            eventName: event.type,
            pointerId: typeof event.pointerId === "number" ? event.pointerId : 0,
            clientX: point.clientX,
            clientY: point.clientY,
            pageX: point.pageX,
            pageY: point.pageY,
            key: event.key ?? null,
            ctrlKey: event.ctrlKey === true,
            shiftKey: event.shiftKey === true,
            altKey: event.altKey === true,
            metaKey: event.metaKey === true,
            matchedSelector: matched ? subscription.selector : null,
            contextElementSelector: getElementSelector(contextElement),
            targetTagName: target?.tagName?.toLowerCase?.() ?? null,
            targetId: target?.id ?? null,
            targetClassName: typeof target?.className === "string" ? target.className : null,
        };
    }

    function getEventPoint(event) {
        if (typeof event.clientX === "number" || typeof event.clientY === "number") {
            return {
                clientX: event.clientX ?? 0,
                clientY: event.clientY ?? 0,
                pageX: event.pageX ?? ((event.clientX ?? 0) + window.scrollX),
                pageY: event.pageY ?? ((event.clientY ?? 0) + window.scrollY),
            };
        }

        const touch = event.changedTouches?.[0] ?? event.touches?.[0];

        return {
            clientX: touch?.clientX ?? 0,
            clientY: touch?.clientY ?? 0,
            pageX: touch?.pageX ?? ((touch?.clientX ?? 0) + window.scrollX),
            pageY: touch?.pageY ?? ((touch?.clientY ?? 0) + window.scrollY),
        };
    }

    function getListenerTargetNames(eventName) {
        if (eventName === "blur") {
            return ["window"];
        }

        return ["document"];
    }

    function getListenerTarget(targetName) {
        return targetName === "window" ? window : document;
    }

    function createListenerKey(eventName, capture, targetName) {
        return `${eventName}:${capture ? "1" : "0"}:${targetName}`;
    }

    function getElementSelector(element) {
        if (!(element instanceof Element)) {
            return null;
        }

        if (element.id) {
            return `[id="${escapeCssStringValue(element.id)}"]`;
        }

        const attributeName = "data-b-document-observer-context-id";
        let contextId = element.getAttribute(attributeName);

        if (!contextId) {
            contextId = `b-document-observer-context-${++nextContextElementId}`;
            element.setAttribute(attributeName, contextId);
        }

        return `[${attributeName}="${escapeCssStringValue(contextId)}"]`;
    }

    function escapeCssStringValue(value) {
        return String(value).replace(/\\/g, "\\\\").replace(/"/g, "\\\"");
    }
}

function getDocumentEventType(eventName) {
    switch (eventName) {
        case "pointerdown":
            return 1;
        case "pointermove":
            return 2;
        case "pointerup":
            return 3;
        case "pointercancel":
            return 4;
        case "mousedown":
            return 5;
        case "mousemove":
            return 6;
        case "mouseup":
            return 7;
        case "touchmove":
            return 8;
        case "touchend":
            return 9;
        case "touchcancel":
            return 10;
        case "click":
            return 11;
        case "dblclick":
            return 12;
        case "keydown":
            return 13;
        case "keyup":
            return 14;
        case "focusin":
            return 15;
        case "focusout":
            return 16;
        case "dragstart":
            return 17;
        case "dragover":
            return 18;
        case "dragend":
            return 19;
        case "drop":
            return 20;
        case "contextmenu":
            return 21;
        case "blur":
            return 22;
        default:
            return 0;
    }
}